_posts/2016-02-24-publishing-data-with-odata-feeds.md

```diff
@@ -0,0 +1,100 @@
+  }
+}
+```
+Connection strings for `mongoDatabase` were configured with `secondaryPreferred` option which basically channeled requests to secondary members in a mongodb server cluster if available. That way we take advantage of our replication strategy and off load OData read requests to the databases used less from our system.
+
+Repositories like that were injected into `ODataControllers`:
+```csharp
+public class SomeEntityController : ODataController
+{
+    private readonly IEntityDataStore _store;
+
+    public AuditorReportController(IEntityDataStore store)
+    {
+        _store = store;
+    }
+
+    [EnableQuery]
+    public IHttpActionResult Get()
+    {
+        return Ok(_store.GetAll());
+    }
+
+    [EnableQuery]
+    public IHttpActionResult Get([FromODataUri] string key)
+    {
+        return Ok(_store.GetAll().Where(entity => entity.Id == key));
+    }
+}
+```
+So, OData controllers supporting only HTTP GET requests expose the IQueryable source of some entity rendering them a readonly endpoint to that data source. An extra single result method was required to fetch entities by their id which is passed through the resource path. The `[EnableQuery]` marks the controller suitable to handle complex OData query expressions which are being passed from the client to the controller using querystring parameters. Thus, by exposing an IQueryable entity through an OData endpoint, query building becomes a responsibility of the client able to synthesize the OData query using expressions like `$filter`, `$select`, `$top`, `$skip` etc.
+
+Finally, defining the routes. Our OWIN Startup class had nothing special, using just the `ODataConvetionModelBuilder` to describe our POCO classes to an Entity Data Model (EDM):
+```csharp
+public class Startup
+{
+  public void Configuration(IAppBuilder app)
+  {
+     var config = new HttpConfiguration();
+
+     var builder = new ODataConventionModelBuilder();
+
+     builder.EntitySet<SomeEntity>("SomeEntities");
+     builder.EntitySet<SomeOtherEntity>("SomeOtherEntities");
+     builder.EntitySet<AwesomeEntity>("AwesomeEntities");
+
+     config.MapODataServiceRoute(
+      routeName: "EntityRoute",
+      routePrefix: "readOnlyOdata",
+      model: builder.GetEdmModel());
+
+     app.UseWebApi(config);
+  }
+}
+```
+Every `EntitySet` defined, map a route to their corresponding controllers. The EntitySet name is used as a resource path in the OData uri. Details about all endpoints uri can be found in the service document. If `http://odatafeed` is the address for our WebAPI project, hitting the service root `http://odatafeed/readOnlyOdata` will reply back with the service document. A JSON object listing the urls of the endpoints available e.g. `http://odatafeed/readOnlyOdata/SomeEntities` etc. Everything that follows after the service root up to querystring is the resource path used to navigate through the hierarchies defined in the EDM.
```

Comment:
> Every `EntitySet` defined, map a route to their corresponding controllers

Every call to `EntitySet` defines a map between a route and its corresponding controller.

--------
>OData uri

OData URI

--------
> Details about all endpoints uri can be found in the service document.

Details of all available endpoint URIs can be found in the service document.